daceml/onnx/op_implementations/fpga_implementations.py

```diff
@@ -0,0 +1,2600 @@
+                "read_weights",
+                {
+                    "b": "0:{}".format(
+                        batch_size
+                    ),  # the batch map loops over every image in the batch
+                    "n0": "0:{}/{}".format(num_filters, P),
+                    "cin": "0:{}".format(num_channels),
+                    "hx": "0:{}".format(filter_hx),
+                    "hy": "0:{}".format(filter_hy)
+                },
+                schedule=dace.ScheduleType.FPGA_Device)
+
+            # use a different map, and unroll it if necessary (otherwise reading weights will slow down everythin)
+            unroll_inner_map = P > (M + L) and P <= 16
+            send_map_entry, send_map_exit = state.add_map(
+                "send_weights", {"n1": "0:{}".format(P)},
+                schedule=dace.ScheduleType.FPGA_Device,
+                unroll=unroll_inner_map)
+
+            mem = state.add_read("W")
+            pipe = state.add_write("W_pipe")
+            tasklet = state.add_tasklet("read_W", {"from_memory"},
+                                        {"to_kernel"},
+                                        "to_kernel = from_memory")
+
+            state.add_memlet_path(
+                mem,
+                entry,
+                send_map_entry,
+                tasklet,
+                dst_conn="from_memory",
+                memlet=dace.Memlet("W[n0 * {} + n1, cin, hx, hy]".format(P)))
+            state.add_memlet_path(tasklet,
+                                  send_map_exit,
+                                  exit,
+                                  pipe,
+                                  src_conn="to_kernel",
+                                  memlet=dace.Memlet(
+                                      "W_pipe[{} -n1 -1]".format(P)))
+
+        def make_read_im2col(state, sdfg, vec_width=1):
+
+            # Matrix B will be the im2col matrix. We will build it row-by-row
+            # to facilitate streaming in the systolic MMM, avoiding storing it back to memory
+            # Note: this will require to load multiple times the input feature, yet this save I/Os
+            # The im2col matrix has size (num_channels * filter_hx * filter_hy) x (output_size_y * output_size_x)
+
+            # gear boxing: we read plain data types, we stream vector data types
+            # Therefore we have two maps, the innermost is unrolled
+            im2col_me, im2col_mx = state.add_map(
+                "im2col_map",
+                {
+                    "b": "0:{}".format(batch_size),
+                    "n": "0:{}/{}".format(
+                        num_filters, P),  # repeat B for computing the result
+                    "cin": "0:{}".format(num_channels),
+                    "hx": "0:{}".format(filter_hx),
+                    "hy": "0:{}".format(filter_hy),
+                    "x": "0:{}".format(output_size_x),
+                    "y0": "0:{}/{}".format(output_size_x, vec_width),
```

Comment:
```suggestion
                    "y0": "0:{}/{}".format(output_size_y),
```
I think `output_size_y` already accounts for vectorization; if `output_size_x` is used I think we should add a test to verify the input is square-shaped.